### DISCUSSION TITLE
Dynamic dependencies for Paper Plugin using bootstrapper

### DISCUSSION THREAD

**Discussion by raikasdev** [PaperMC/Paper#10574] [Ideas]
### Is your feature request related to a problem?

I maintain a plugin called [CraftJS](https://github.com/raikasdev/CraftJS), which currently has a problem with dependencies of the CraftJS-plugins. Currently I haven't released the plugin for broaded use, so I have just added dependencies I need to the CraftJS paper-plugin.yml

### Describe the solution you'd like.

Allow bootstrapper code to dynamically add dependencies, soft-dependencies and plugin load-order information.

### Describe alternatives you've considered.

~~Overwriting the getPluginMeta function could maybe work, but I think this could be an good use case for bootstrappers.~~
It was a final method.

**Comment by olijeffers0n** [comment]
I don’t have a huge knowledge of the dependency system, but I would imagine this has the potential to cause huge issues with regards to cyclical and load-order problems.

**Comment by Owen1212055** [comment]
Use the `has-open-classloader: true` flag in the paper-plugin.yml

In general I'm not too keen on allowing things to be dynamic that are registered in the yml. There's a chance that something like this may be implemented in the future, but, at least for now I'd rather keep this static.

The above flag should allow you to access all classloaders for other plugins, so you shouldn't have resolution errors for the time being.

**Comment by raikasdev** [comment]
The thing wasn't certainly just the classloading, but also that the plugin also should be loaded after all other plugins. Is this possible to do dynamically?

**Comment by Owen1212055** [comment]
Yeah, it's defo possible, just not sure to what extent here this should be done. I'll reopen this for now.

**Comment by emilyy-dev** [comment]
I feel like allowing to declare load order dynamically makes it too easy to accidentally introduce cyclic load orders if not careful, it's easy enough as it is with the "right" plugins (B declares load-after A and load-before C, C declares load-before A).
The intent of "loading after all other plugins" is not unheard of, but unrealistic. Say both plugin A and B want to load after all plugins have loaded, so plugin A dynamically declares load-after B, and plugin B dynamically declares load-after A. Sure, it's something to report to the respective authors, but if anything, doing this makes it far easier to mess up.
